### ISSUE TITLE
[ILM] Shrink action fails with a single primary shard in conjunction with `max_primary_shard_size`

### ISSUE THREAD

**Issue by fatmcgav** [elastic/elasticsearch#80180]
**Elasticsearch version** (`bin/elasticsearch --version`): `7.14.2`

**Plugins installed**: []

**JVM version** (`java -version`): ESS

**OS version** (`uname -a` if on a Unix-like system): ESS

**Description of the problem including expected versus actual behavior**:

The `shrink` ILM phase failed when trying to apply to an index with a single primary shard and a phase definition that utilises the `max_primary_shard_size` option.

Exception recorded:
```
java.lang.IllegalArgumentException: can't shrink an index with only one shard
	at org.elasticsearch.cluster.metadata.MetadataCreateIndexService.validateShrinkIndex(MetadataCreateIndexService.java:1147)
	at org.elasticsearch.cluster.metadata.MetadataCreateIndexService.prepareResizeIndexSettings(MetadataCreateIndexService.java:1233)
	at org.elasticsearch.cluster.metadata.MetadataCreateIndexService.aggregateIndexSettings(MetadataCreateIndexService.java:809)
	at org.elasticsearch.cluster.metadata.MetadataCreateIndexService.applyCreateIndexRequestWithExistingMetadata(MetadataCreateIndexService.java:631)
	at org.elasticsearch.cluster.metadata.MetadataCreateIndexService.applyCreateIndexRequest(MetadataCreateIndexService.java:334)
	at org.elasticsearch.cluster.metadata.MetadataCreateIndexService.applyCreateIndexRequest(MetadataCreateIndexService.java:377)
	at org.elasticsearch.cluster.metadata.MetadataCreateIndexService$1.execute(MetadataCreateIndexService.java:294)
	at org.elasticsearch.cluster.ClusterStateUpdateTask.execute(ClusterStateUpdateTask.java:48)
	at org.elasticsearch.cluster.service.MasterService.executeTasks(MasterService.java:691)
	at org.elasticsearch.cluster.service.MasterService.calculateTaskOutputs(MasterService.java:313)
	at org.elasticsearch.cluster.service.MasterService.runTasks(MasterService.java:208)
	at org.elasticsearch.cluster.service.MasterService.access$000(MasterService.java:62)
	at org.elasticsearch.cluster.service.MasterService$Batcher.run(MasterService.java:140)
	at org.elasticsearch.cluster.service.TaskBatcher.runIfNotProcessed(TaskBatcher.java:139)
	at org.elasticsearch.cluster.service.TaskBatcher$BatchedTask.run(TaskBatcher.java:177)
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:673)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:241)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:204)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:831)
```

**Steps to reproduce**:

Please include a *minimal* but *complete* recreation of the problem,
including (e.g.) index creation, mappings, settings, query etc.  The easier
you make for us to reproduce it, the more likely that somebody will take the
time to look at it.

 1. Create an index with a single primary shard.
 2. Define an ILM policy with the following `warm` phase:
  ```json
{
  "policy": "metricbeat",
  "phase_definition": {
    "min_age": "0d",
    "actions": {
      "allocate": {
        "number_of_replicas": 1,
        "include": {},
        "exclude": {},
        "require": {}
      },
      "forcemerge": {
        "max_num_segments": 1
      },
      "set_priority": {
        "priority": 25
      },
      "shrink": {
        "max_primary_shard_size": "50gb"
      }
    }
  },
  "version": 18,
  "modified_date_in_millis": 1635781185067
}
  ```
 3. Associate ILM policy with index
 4. Trigger rollover of index.

**Provide logs (if relevant)**:

Related: #33275

**Comment by fatmcgav** [comment]
We have had another occurence of this error and I've captured the ILM `explain` output that was requested:

```
{
  "indices" : {
    "filebeat-7.14.2-2021.10.31-000008" : {
      "index" : "filebeat-7.14.2-2021.10.31-000008",
      "managed" : true,
      "policy" : "filebeat",
      "lifecycle_date_millis" : 1635894729560,
      "age" : "12.33h",
      "phase" : "warm",
      "phase_time_millis" : 1635938605386,
      "action" : "shrink",
      "action_time_millis" : 1635898087540,
      "step" : "ERROR",
      "step_time_millis" : 1635938605539,
      "failed_step" : "shrink",
      "is_auto_retryable_error" : true,
      "failed_step_retry_count" : 65,
      "shrink_index_name" : "shrink-r7yz-filebeat-7.14.2-2021.10.31-000008",
      "step_info" : {
        "type" : "illegal_argument_exception",
        "reason" : "can't shrink an index with only one shard",
        "stack_trace" : """java.lang.IllegalArgumentException: can't shrink an index with only one shard
	at org.elasticsearch.cluster.metadata.MetadataCreateIndexService.validateShrinkIndex(MetadataCreateIndexService.java:1147)
	at org.elasticsearch.cluster.metadata.MetadataCreateIndexService.prepareResizeIndexSettings(MetadataCreateIndexService.java:1233)
	at org.elasticsearch.cluster.metadata.MetadataCreateIndexService.aggregateIndexSettings(MetadataCreateIndexService.java:809)
	at org.elasticsearch.cluster.metadata.MetadataCreateIndexService.applyCreateIndexRequestWithExistingMetadata(MetadataCreateIndexService.java:631)
	at org.elasticsearch.cluster.metadata.MetadataCreateIndexService.applyCreateIndexRequest(MetadataCreateIndexService.java:334)
	at org.elasticsearch.cluster.metadata.MetadataCreateIndexService.applyCreateIndexRequest(MetadataCreateIndexService.java:377)
	at org.elasticsearch.cluster.metadata.MetadataCreateIndexService$1.execute(MetadataCreateIndexService.java:294)
	at org.elasticsearch.cluster.ClusterStateUpdateTask.execute(ClusterStateUpdateTask.java:48)
	at org.elasticsearch.cluster.service.MasterService.executeTasks(MasterService.java:691)
	at org.elasticsearch.cluster.service.MasterService.calculateTaskOutputs(MasterService.java:313)
	at org.elasticsearch.cluster.service.MasterService.runTasks(MasterService.java:208)
	at org.elasticsearch.cluster.service.MasterService.access$000(MasterService.java:62)
	at org.elasticsearch.cluster.service.MasterService$Batcher.run(MasterService.java:140)
	at org.elasticsearch.cluster.service.TaskBatcher.runIfNotProcessed(TaskBatcher.java:139)
	at org.elasticsearch.cluster.service.TaskBatcher$BatchedTask.run(TaskBatcher.java:177)
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:673)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:241)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:204)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:831)
"""
      },
      "phase_execution" : {
        "policy" : "filebeat",
        "phase_definition" : {
          "min_age" : "0d",
          "actions" : {
            "allocate" : {
              "number_of_replicas" : 1,
              "include" : { },
              "exclude" : { },
              "require" : { }
            },
            "forcemerge" : {
              "max_num_segments" : 1
            },
            "set_priority" : {
              "priority" : 25
            },
            "shrink" : {
              "max_primary_shard_size" : "50gb"
            }
          }
        },
        "version" : 21,
        "modified_date_in_millis" : 1635850735993
      }
    }
  }
}
```

**Comment by gaobinlong** [comment]
@fatmcgav , it's useless if you shrink an index with only one shard, so the `Shrink Action` throws the `illegal_argument_exception `, maybe it's not a problem?

**Comment by fatmcgav** [comment]
> @fatmcgav , it's useless if you shrink an index with only one shard, so the `Shrink Action` throws the `illegal_argument_exception `, maybe it's not a problem?

Agreed that technically, it's useless to try and shrink and index with only one shart. 
However the fact that Elasticsearch reports this as an error condition on the index is also redundant in my opinion.

This is inline with the logic applied in #33275. The only difference here is that I'm using `max_primary_shard_size` rather than a specific value of `1`...

**Comment by andreidan** [comment]
Shrink operations can be executed based on replicas however, when `max_primary_shard_size` yields the same number of shards for the target (shrunk) index as the source index has the `ResizeAllocationDecider` will kick in thinking we're performing a clone. A clone operation is executed from primaries only so this could block the shrink operation.

We should tackle the case when `max_primary_shard_size` would yield the same number of shards as the source index already has, and skip the shrink action.

**Comment by ehinkle27** [comment]
was this fixed in a later version was running 8.1 and had same issue.

ava.lang.IllegalArgumentException: can't shrink an index with only one shard
	at org.elasticsearch.cluster.metadata.MetadataCreateIndexService.validateShrinkIndex(MetadataCreateIndexService.java:1324)
	at org.elasticsearch.cluster.metadata.MetadataCreateIndexService.prepareResizeIndexSettings(MetadataCreateIndexService.java:1423)
	at org.elasticsearch.cluster.metadata.MetadataCreateIndexService.aggregateIndexSettings(MetadataCreateIndexService.java:951)
	at org.elasticsearch.cluster.metadata.MetadataCreateIndexService.applyCreateIndexRequestWithExistingMetadata(MetadataCreateIndexService.java:742)
	at org.elasticsearch.cluster.metadata.MetadataCreateIndexService.applyCreateIndexRequest(MetadataCreateIndexService.java:341)
	at org.elasticsearch.cluster.metadata.MetadataCreateIndexService.applyCreateIndexRequest(MetadataCreateIndexService.java:395)
	at org.elasticsearch.cluster.metadata.MetadataCreateIndexService$1.execute(MetadataCreateIndexService.java:293)
	at org.elasticsearch.cluster.ClusterStateTaskExecutor$1.execute(ClusterStateTaskExecutor.java:142)
	at org.elasticsearch.cluster.service.MasterService.executeTasks(MasterService.java:818)
	at org.elasticsearch.cluster.service.MasterService.calculateTaskOutputs(MasterService.java:397)
	at org.elasticsearch.cluster.service.MasterService.runTasks(MasterService.java:237)
	at org.elasticsearch.cluster.service.MasterService$Batcher.run(MasterService.java:162)

**Comment by dakrone** [comment]
@ehinkle27 this was fixed in 8.2.0+

**Comment by tuckerpm** [comment]
Any chance this made 7.x?

**Comment by dakrone** [comment]
@tuckerpm this was a fairly substantial change and was not backported to 7.x.

**Comment by fschoening** [comment]
Should you have an index stuck in this error state, you can manually move it to the next phase using this snippet:
```
POST _ilm/move/my_index
{
  "current_step": { 
    "phase": "warm",
    "action": "shrink",
    "name": "ERROR"
  },
  "next_step": { 
    "phase": "warm",
    "action": "forcemerge", 
    "name": "forcemerge" 
  }
}
```

Of course assuming the next phase is forcemerge. If in doubt, investigate using `GET my_index/_ilm/explain`

**Comment by vppfy** [comment]
Encountered this issue unexpectedly on 8.12.2. 400+ indices got this error, kicking them manually through the stuck step is not an option.

**Comment by dakrone** [comment]
@vppfy this should be fixed in 8.2.0+, can you share the failure details and how you determined it was the same issue?
